funtofem/mphys/mphys_meld.py

```diff
@@ -269,89 +316,206 @@ def compute_jacvec_product(self, inputs, d_inputs, d_outputs, mode):
-                        pass
-                    else:
-                        raise ValueError("forward mode requested but not implemented")
-
-        if mode == "rev":
-            if "f_struct" in d_outputs:
-                d_out = np.zeros(self.struct_nnodes * 3, dtype=TransferScheme.dtype)
-                for i in range(3):
-                    d_out[i::3] = d_outputs["f_struct"][i :: self.struct_ndof]
-
-                if "u_struct" in d_inputs:
-                    d_in = np.zeros(self.struct_nnodes * 3, dtype=TransferScheme.dtype)
-                    # df_s/du_s^T * psi = - dL/du_s^T * psi
-                    self.meld.applydLduSTrans(d_out, d_in)
+                    if "x_aero0" in d_inputs:
+                        # df_s/dx_a0^T * psi = - psi^T * dL/dx_a0 in F2F terminology
+                        prod = np.zeros(
+                            len(body.aero_coord_indices), dtype=TransferScheme.dtype
+                        )
+                        body.meld.applydLdxA0(d_out, prod)
+                        d_inputs["x_aero0"][body.aero_coord_indices] -= np.array(
+                            prod, dtype=float
+                        )
 
-                    for i in range(3):
-                        d_inputs["u_struct"][i :: self.struct_ndof] -= np.array(
-                            d_in[i::3], dtype=float
+                    if "x_struct0" in d_inputs:
+                        # df_s/dx_s0^T * psi = - psi^T * dL/dx_s0 in F2F terminology
+                        prod = np.zeros(
+                            len(body.struct_coord_indices), dtype=TransferScheme.dtype
+                        )
+                        body.meld.applydLdxS0(d_out, prod)
+                        d_inputs["x_struct0"][body.struct_coord_indices] -= np.array(
+                            prod, dtype=float
                         )
 
-                if "f_aero" in d_inputs:
-                    # df_s/df_a^T psi = - dL/df_a^T * psi = -dD/du_s * psi
-                    prod = np.zeros(self.aero_nnodes * 3, dtype=TransferScheme.dtype)
-                    self.meld.applydDduS(d_out, prod)
-                    d_inputs["f_aero"] -= np.array(prod, dtype=float)
 
-                if "x_aero0" in d_inputs:
-                    # df_s/dx_a0^T * psi = - psi^T * dL/dx_a0 in F2F terminology
-                    prod = np.zeros(self.aero_nnodes * 3, dtype=TransferScheme.dtype)
-                    self.meld.applydLdxA0(d_out, prod)
-                    d_inputs["x_aero0"] -= np.array(prod, dtype=float)
+class MeldBodyInstance:
+    """
+    Class that helps split OpenMDAO input/output indices for multiple bodies,
+    with a separate MELD instance for each
+    """
+
+    def __init__(
+        self,
+        comm,
+        isym=-1,
+        n=200,
+        beta=0.5,
+        aero_gid=None,  # list of aero grid IDs to couple
+        aero_nnodes=None,
+        struct_gid=None,  # list of struct grid IDs to couple
```

Comment:
Would it be more correct to change `gid` to `tags` throughout the code to more accurately reflect the changes made to https://github.com/OpenMDAO/mphys/pull/158 ?

Reply:
Forgot to respond to this... to me, `tags` sounds like a group of nodes corresponding to a boundary, property group, etc. I'm open to name changes, but would prefer something relating to `node` ids rather than `tag` ids?

Reply:
Sorry, this again is me misunderstanding the code, I thought this input was the tags list, but now I see that the `MeldBuilder` deals with those and these are the actual node indices. `node_ids` would be slightly clearer.